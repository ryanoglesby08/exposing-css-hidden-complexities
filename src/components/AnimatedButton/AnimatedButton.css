.button {
  padding: 1rem;

  font-size: 1rem;
  color: #ffffff;

  cursor: pointer;
}

.inline {
  display: inline-block;
  min-width: 200px;
}

.fullWidth {
  display: block;
  width: 100%;
}
