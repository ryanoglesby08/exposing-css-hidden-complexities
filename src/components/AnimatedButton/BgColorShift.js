import React from "react";
import { number, string } from "prop-types";
import { Transition } from "react-transition-group";

const defaultStyle = (startBgColor, timeout) => ({
  transition: `background-color ${timeout}ms`,
  backgroundColor: startBgColor
});

const transitionStyles = endBgColor => {
  const styles = {
    backgroundColor: endBgColor
  };

  return {
    entering: styles,
    entered: styles
  };
};

const BgColorShift = ({ start, end, children, ...rest }) => (
  <Transition {...rest}>
    {status => {
      const styles = {
        ...defaultStyle(start, rest.timeout),
        ...transitionStyles(end)[status]
      };

      return children(styles);
    }}
  </Transition>
);
BgColorShift.propTypes = {
  timeout: number.isRequired,
  start: string.isRequired,
  end: string.isRequired
};

export default BgColorShift;
