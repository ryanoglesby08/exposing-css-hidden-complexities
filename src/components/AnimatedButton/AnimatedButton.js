import React, { Component } from "react";
import Media from "react-media";

import BgColorShift from "./BgColorShift";

import styles from "./AnimatedButton.css";
const smallViewportMaxWidth = 767;
const bgColors = {
  default: "#2892d7",
  hover: "#1d70a2"
};

class Button extends Component {
  state = {
    hover: false
  };

  toggleHover = () => {
    this.setState({ hover: !this.state.hover });
  };

  render() {
    const { className, children } = this.props;

    return (
      <BgColorShift
        timeout={500}
        in={this.state.hover}
        start={bgColors.default}
        end={bgColors.hover}
      >
        {bgColorStyles => (
          <button
            onMouseEnter={this.toggleHover}
            onMouseLeave={this.toggleHover}
            style={bgColorStyles}
            className={`${styles.button} ${className}`}
          >
            {children}
          </button>
        )}
      </BgColorShift>
    );
  }
}

export default ({ children }) => (
  <Media query={{ maxWidth: smallViewportMaxWidth }}>
    {matches =>
      matches ? (
        <Button className={styles.fullWidth}>{children}</Button>
      ) : (
        <Button className={styles.inline}>{children}</Button>
      )
    }
  </Media>
);
