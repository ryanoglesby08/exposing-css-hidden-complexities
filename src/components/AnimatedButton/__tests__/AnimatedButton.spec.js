import React from "react";
import { mount, render } from "enzyme";

import mockMatchMedia from "../../../__mocks__/matchMedia";

import AnimatedButton from "../AnimatedButton";

describe("AnimatedButton", () => {
  beforeEach(() => {
    mockMatchMedia();
  });

  it("renders pre color shift", () => {
    const button = render(<AnimatedButton>Press me</AnimatedButton>);

    expect(button).toMatchSnapshot();
  });

  it("renders post color shift", () => {
    const button = mount(<AnimatedButton>Press me</AnimatedButton>);

    button.simulate("mouseEnter");

    expect(button).toMatchSnapshot();
  });
});
