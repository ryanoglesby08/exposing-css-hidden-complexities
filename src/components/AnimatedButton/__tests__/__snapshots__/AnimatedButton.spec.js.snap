// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedButton renders post color shift 1`] = `
<Component>
  <Media
    defaultMatches={true}
    query={
      Object {
        "maxWidth": 767,
      }
    }
  >
    <Button
      className="fullWidth"
    >
      <BgColorShift
        end="#1d70a2"
        in={true}
        start="#2892d7"
        timeout={500}
      >
        <Transition
          appear={false}
          enter={true}
          exit={true}
          in={true}
          mountOnEnter={false}
          onEnter={[Function]}
          onEntered={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onExiting={[Function]}
          timeout={500}
          unmountOnExit={false}
        >
          <button
            className="button fullWidth"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "backgroundColor": "#1d70a2",
                "transition": "background-color 500ms",
              }
            }
          >
            Press me
          </button>
        </Transition>
      </BgColorShift>
    </Button>
  </Media>
</Component>
`;

exports[`AnimatedButton renders pre color shift 1`] = `
<button
  class="button fullWidth"
  style="transition:background-color 500ms;background-color:#2892d7"
>
  Press me
</button>
`;
